File: backend-py/app/api/query.py
"""Query-string binding that mirrors MVC's model binder.

Conversion failures are collected rather than raised one at a time, so a
request with three bad values reports all three. An *absent* key is not the
same as an empty one: "?month=" is a failure, not "this month".
"""

from __future__ import annotations

import uuid
from datetime import datetime

from fastapi import Request

from app.core.errors import ValidationError
from app.domain import validation as rules
from app.domain.money import parse_wire_date


class QueryReader:
    def __init__(self, request: Request) -> None:
        self._params = request.query_params
        self._errs = ValidationError()

    def text(self, key: str) -> str | None:
        """None when the caller omitted the key entirely."""
        return self._params.get(key) if key in self._params else None

    def number(self, key: str, field: str) -> int | None:
        raw = self.text(key)
        if raw is None:
            return None
        try:
            return int(raw)
        except ValueError:
            rules.invalid_value(self._errs, raw, field)
            return None

    def number_or(self, key: str, field: str, fallback: int) -> int:
        parsed = self.number(key, field)
        return fallback if parsed is None else parsed

    def identifier(self, key: str, field: str) -> uuid.UUID | None:
        raw = self.text(key)
        if raw is None:
            return None
        try:
            return uuid.UUID(raw)
        except ValueError:
            rules.invalid_value(self._errs, raw, field)
            return None

    def moment(self, key: str, field: str) -> datetime | None:
        raw = self.text(key)
        if raw is None:
            return None
        parsed = parse_wire_date(raw)
        if parsed is None:
            rules.invalid_value(self._errs, raw, field)
        return parsed

    def enum(self, key: str, field: str, kind: type) -> object | None:
        raw = self.text(key)
        if raw is None:
            return None
        parsed = kind.parse(raw)  # type: ignore[attr-defined]
        if parsed is None:
            rules.invalid_value(self._errs, raw, field)
        return parsed

    def done(self) -> None:
        """Raises the accumulated conversion failures, if any."""
        rules.raise_if_any(self._errs)

File: backend-node/src/api/query.ts
/**
 * Query-string binding that mirrors MVC's model binder.
 *
 * Conversion failures are collected rather than raised one at a time, so a
 * request with three bad values reports all three. An *absent* key is not the
 * same as an empty one: "?month=" is a failure, not "this month".
 */

import type { Request } from 'express'
import { FieldErrors } from '../core/errors.js'
import { type EnumKind, type EnumValue, parseEnum } from '../domain/enums.js'
import { type Instant, parseWireDate } from '../domain/instant.js'
import { invalidValueMessage } from '../domain/validation.js'

const UUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i

export class QueryReader {
  private readonly params: Record<string, unknown>
  private readonly errs = new FieldErrors()

  constructor(request: Request) {
    this.params = request.query
  }

  /** null when the caller omitted the key entirely. */
  text(key: string): string | null {
    const value = this.params[key]
    if (value === undefined) return null
    // A repeated key arrives as an array; the binder reads the first one.
    if (Array.isArray(value)) return typeof value[0] === 'string' ? value[0] : ''
    return typeof value === 'string' ? value : ''
  }

  number(key: string, field: string): number | null {
    const raw = this.text(key)
    if (raw === null) return null

    // Deliberately strict: MVC's binder rejects "12abc" rather than reading 12.
    if (!/^[+-]?\d+$/.test(raw.trim())) {
      this.errs.add(field, invalidValueMessage(raw, field))
      return null
    }
    return Number(raw)
  }

  numberOr(key: string, field: string, fallback: number): number {
    const parsed = this.number(key, field)
    return parsed ?? fallback
  }

  identifier(key: string, field: string): string | null {
    const raw = this.text(key)
    if (raw === null) return null
    if (!UUID_PATTERN.test(raw)) {
      this.errs.add(field, invalidValueMessage(raw, field))
      return null
    }
    return raw
  }

  moment(key: string, field: string): Instant | null {
    const raw = this.text(key)
    if (raw === null) return null

    const parsed = parseWireDate(raw)
    if (parsed === null) this.errs.add(field, invalidValueMessage(raw, field))
    return parsed
  }

  enum(key: string, field: string, kind: EnumKind): EnumValue | null {
    const raw = this.text(key)
    if (raw === null) return null

    const parsed = parseEnum(kind, raw)
    if (parsed === null) this.errs.add(field, invalidValueMessage(raw, field))
    return parsed
  }

  /** Throws the accumulated conversion failures, if any. */
  done(): void {
    this.errs.raiseIfAny()
  }
}
